File: study/lua/src/app/ui/component/CommonJadeIcon.lua
--
-- Author: hedl
-- Date: 2017-02-22 18:02:15
-- 装备Icon

local CommonIconBase = import(".CommonIconBase")
local CommonJadeIcon = class("CommonJadeIcon", CommonIconBase)
local UIHelper = require("yoka.utils.UIHelper")
local ComponentIconHelper = require("app.ui.component.ComponentIconHelper")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local EffectGfxNode = require("app.effect.EffectGfxNode")
local PopupJadeDetail = require("app.scene.view.equipmentJade.PopupJadeDetail")

local EXPORTED_METHODS = {
    "setSysId",
    "setIconScale"
}

function CommonJadeIcon:ctor()
    CommonJadeIcon.super.ctor(self)
    self._textItemTopNum = nil -- 顶部数字按钮
    self._equipId = nil
    self._sysId = 0
    self._type = TypeConvertHelper.TYPE_JADE_STONE
    self._effect1 = nil
    self._effect2 = nil
end

function CommonJadeIcon:_init()
    CommonJadeIcon.super._init(self)
    self._nodeEffectDown = ccui.Helper:seekNodeByName(self._target, "NodeEffectDown")
    self._nodeEffectUp = ccui.Helper:seekNodeByName(self._target, "NodeEffectUp")
    self:setTouchEnabled(true)
end

function CommonJadeIcon:bind(target)
    CommonJadeIcon.super.bind(self, target)

    cc.setmethods(target, self, EXPORTED_METHODS)
end

function CommonJadeIcon:unbind(target)
    CommonJadeIcon.super.unbind(self, target)

    cc.unsetmethods(target, EXPORTED_METHODS)
end

function CommonJadeIcon:setSysId(sysId)
    self._sysId = sysId
end

--根据传入参数，创建并，更新UI
function CommonJadeIcon:updateUI(value, size)
    self._sysId = value
    local itemParams = CommonJadeIcon.super.updateUI(self, value, size)

    if itemParams.size ~= nil then
        self:setCount(itemParams.size)
    end
    self:showIconEffect()
end

function CommonJadeIcon:_onTouchCallBack(sender, state)
    -----------防止拖动的时候触发点击
    if (state == ccui.TouchEventType.ended) then
        local moveOffsetX = math.abs(sender:getTouchEndPosition().x - sender:getTouchBeganPosition().x)
        local moveOffsetY = math.abs(sender:getTouchEndPosition().y - sender:getTouchBeganPosition().y)
        if moveOffsetX < 20 and moveOffsetY < 20 then
            if self._callback then
                self._callback(self._target, self._itemParams)
            else
                local popupJadeDetail = PopupJadeDetail.new(TypeConvertHelper.TYPE_JADE_STONE, self._itemParams.cfg.id)
                popupJadeDetail:openWithAction()
            end
        end
    end
end

function CommonJadeIcon:setIconScale(scale)
    self._imageIcon:setScale(scale)
end

function CommonJadeIcon:removeLightEffect()
end

function CommonJadeIcon:showIconEffect(scale)
end

return CommonJadeIcon

File: study/ts/script/ui/component/CommonJadeIcon.ts
import CommonIconBase from "./CommonIconBase";
import { TypeConvertHelper } from "../../utils/TypeConvertHelper";
import PopupJadeDetail from "../../scene/view/equipmentJade/PopupJadeDetail";

const { ccclass, property } = cc._decorator;

@ccclass
export default class CommonJadeIcon extends CommonIconBase {

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeEffectDown: cc.Node = null;


    @property({
        type: cc.Node,
        visible: true
    })
    _nodeEffectUp: cc.Node = null;

    _textItemTopNum = null;
    _equipId = null;
    _sysId = 0;
    _type = TypeConvertHelper.TYPE_JADE_STONE;
    _effect1 = null;
    _effect2 = null;

    private _isClick:boolean = true;

    constructor() {
        super();
        this._type = TypeConvertHelper.TYPE_JADE_STONE;
    }

    onLoad() {
        super.onLoad();
        this._nodeEffectDown = this._nodeEffectDown || this.node.getChildByName('NodeEffectDown');
        this._nodeEffectUp = this._nodeEffectUp || this.node.getChildByName('NodeEffectUp');
    }

    setSysId(sysId) {
        this._sysId = sysId;
    }

    updateUI(value, size?) {
        this._sysId = value;
        var itemParams = super.updateUI(value, size);
        if (itemParams.size != null) {
            this.setCount(itemParams.size);
        }
        this.showIconEffect();
    }

    public setTouchEnabled(st):void{
        this._isClick = st;
        super.setTouchEnabled(st);
    }

    _onTouchCallBack(sender, state) {

        if(!this._isClick)return;
        if (this._callback) {
            this._callback(this._target, this._itemParams);
        }
        else {
            PopupJadeDetail.getIns(PopupJadeDetail, (p: PopupJadeDetail) => {
                p.ctor(TypeConvertHelper.TYPE_JADE_STONE, this._itemParams.cfg.id);
                p.openWithAction();
            })
        }
    }
    setIconScale(scale) {
        this._imageIcon.node.setScale(scale);
    }
    removeLightEffect() {
    }
    showIconEffect(scale?) {
    }

}